File: c/lib/shared-state.c
#include <malloc.h>
#include "shared-state.h"

SharedState *alloc_shared_state(int bars) {
    SharedState *state = malloc(sizeof(SharedState));
    if (state == NULL) return NULL;

    state->firstBar = 0;

    if ((state->numA    = malloc(sizeof(double) * bars)) == NULL) return NULL;
    if ((state->numB    = malloc(sizeof(double) * bars)) == NULL) return NULL;
    if ((state->numC    = malloc(sizeof(double) * bars)) == NULL) return NULL;
    if ((state->numD    = malloc(sizeof(double) * bars)) == NULL) return NULL;
    if ((state->balance = malloc(sizeof(double) * bars)) == NULL) return NULL;
    if ((state->equity  = malloc(sizeof(double) * bars)) == NULL) return NULL;

    return state;
}

void free_shared_state(SharedState *state) {
    if (state == NULL) return;

    if (state->numA != NULL) {
        free(state->numA);
        state->numA = NULL;
    }
    if (state->numB != NULL) {
        free(state->numB);
        state->numB = NULL;
    }
    if (state->numC != NULL) {
        free(state->numC);
        state->numC = NULL;
    }
    if (state->numD != NULL) {
        free(state->numD);
        state->numD = NULL;
    }
    if (state->balance != NULL) {
        free(state->balance);
        state->balance = NULL;
    }
    if (state->equity != NULL) {
        free(state->equity);
        state->equity = NULL;
    }

    free(state);
    state = NULL;
}

File: fortran/lib/shared-state.f90
module Shared_state_module
    use iso_fortran_env, only: real64
    implicit none
    private

    type :: Shared_state_type
        integer                                 :: firstBar
        real(real64), dimension(:), allocatable :: numA
        real(real64), dimension(:), allocatable :: numB
        real(real64), dimension(:), allocatable :: numC
        real(real64), dimension(:), allocatable :: numD
        real(real64), dimension(:), allocatable :: balance
        real(real64), dimension(:), allocatable :: equity
    end type Shared_state_type

    public Shared_state_type
    public init_shared_state, deinit_shared_state

    contains

    subroutine init_shared_state(state, bars)
        type(Shared_state_type), intent(inout) :: state
        integer,                 intent(in)    :: bars

        state%firstBar = 0
        allocate(state%numA   (bars), source=0.0_real64)
        allocate(state%numB   (bars), source=0.0_real64)
        allocate(state%numC   (bars), source=0.0_real64)
        allocate(state%numD   (bars), source=0.0_real64)
        allocate(state%balance(bars), source=0.0_real64)
        allocate(state%equity (bars), source=0.0_real64)
    end subroutine init_shared_state

    subroutine deinit_shared_state(state)
        type(Shared_state_type), intent(inout) :: state

        deallocate(state%numA   )
        deallocate(state%numB   )
        deallocate(state%numC   )
        deallocate(state%numD   )
        deallocate(state%balance)
        deallocate(state%equity )
    end subroutine deinit_shared_state
end module Shared_state_module
